### PR TITLE
multi trackers working

### PR DESCRIPTION
**Pull request by chrisnojima**
also substores can go over ipc

@keybase/react-hackers

### CODE REVIEW

**Comment by chrisnojima** [comment]
@keybase/react-hackers using the close flag

**Comment by cjb** [comment]
Just tried multiple trackers at once and closing some, worked like a charm.

File: desktop/npm-debug.log [hunk 1]
```diff
@@ -0,0 +1,49 @@
+0 info it worked if it ends with ok
```
Comment (MarcoPolo) [inline desktop/npm-debug.log]: accidental checkin?

File: react-native/react/native/remote-manager.desktop.js [hunk 2]
```diff
@@ -34,29 +34,56 @@ class RemoteManager extends Component {
...
+        popups[username] = (
+          <RemoteComponent
+            windowsOpts={{
+              height: 332,
+              width: 520,
+              frame: false,
+              resizable: false
+            }}
+            waitForState
+            onRemoteClose={this.props.onCloseFromHeader}
```
Comment (MarcoPolo) [inline react-native/react/native/remote-manager.desktop.js]: shouldn't you have a key attribute here so react can determine which one is for which?

File: react-native/react/reducers/tracker.js [hunk 3]
```diff
@@ -29,67 +28,67 @@ type State = {
...
+      followsYou: false,
+      avatar: null,
+      location: '' // TODO: get this information
+    }
   }
 }
 
-export default function (state: State = initialState, action: Action): State {
+// Just mutating the substate of the tracker
+function updateUserState (rootState: State, trackerState: TrackerState, action: Action) {
```
Comment (MarcoPolo) [inline react-native/react/reducers/tracker.js]: What does this function return?

File: react-native/react/native/remote-manager.desktop.js (hunk 2 above)
Comment (chrisnojima) [inline react-native/react/native/remote-manager.desktop.js]: yeah i forgot to add that

File: react-native/react/reducers/tracker.js [hunk 4]
```diff
@@ -108,82 +107,117 @@ export default function (state: State = initialState, action: Action): State {
...
-        return state
+        return rootState
       }
 
       const rp: RemoteProof = action.payload.remoteProof
       const lcr: LinkCheckResult = action.payload.linkCheckResult
       return {
-        ...state,
-        proofs: updateProof(state.proofs, rp, lcr)
+        proofs: updateProof(trackerState.proofs, rp, lcr) || []
```
Comment (MarcoPolo) [inline react-native/react/reducers/tracker.js]: I don't think that [] is necessary?

File: react-native/react/reducers/tracker.js [hunk 5]
```diff
@@ -285,5 +319,5 @@ function updateProof (proofs: Array<Proof>, rp: RemoteProof, lcr: LinkCheckResul
     updated.push(remoteProofToProof(rp, lcr))
   }
 
-  return updated
+  return updated || []
```
Comment (MarcoPolo) [inline react-native/react/reducers/tracker.js]: when is updated ever null?

File: react-native/react/native/remote-manager.desktop.js (hunk 2 above)
Comment (chrisnojima) [inline react-native/react/native/remote-manager.desktop.js]: done

File: react-native/react/reducers/tracker.js [hunk 6]
```diff
@@ -108,82 +107,117 @@ export default function (state: State = initialState, action: Action): State {
...
-        ...state,
-        serverStarted
+        userInfo: action.payload.userInfo
       }
 
     case Constants.markActiveIdentifyUi:
       const serverActive = action.payload && !!action.payload.active || false
       // The server wasn't active and now it is, we reset closed state
-      const closed = (showAllTrackers && !state.serverActive && serverActive) ? false : state.closed
+      const closed = (showAllTrackers && !rootState.serverActive && serverActive) ? false : trackerState.closed
```
Comment (MarcoPolo) [inline react-native/react/reducers/tracker.js]: so there is one server instance per session id. I think this might be broken.

File: react-native/react/reducers/tracker.js [hunk 7]
```diff
@@ -29,67 +28,67 @@ type State = {
...
 }
 
-export default function (state: State = initialState, action: Action): State {
+// Just mutating the substate of the tracker
+function updateUserState (rootState: State, trackerState: TrackerState, action: Action) {
   switch (action.type) {
     case Constants.onFollowChecked:
       if (action.payload == null) {
-        return state
+        return rootState
```
Comment (MarcoPolo) [inline react-native/react/reducers/tracker.js]: why do you return rootState here?

Comment (chrisnojima) [inline react-native/react/reducers/tracker.js]: if this thing does 'nothing' it returns the root state, teh caller detects this and bails vs making duplicate copies. we should re-discuss immutable...

Comment (chrisnojima) [inline react-native/react/reducers/tracker.js]: also this code is logically teh same as it was before, rootState was state, but now we're in a substate

File: react-native/react/reducers/tracker.js [hunk 8]
```diff
@@ -29,64 +28,64 @@ type State = {
...
+      followsYou: false,
+      avatar: null,
+      location: '' // TODO: get this information
+    }
   }
 }
 
-export default function (state: State = initialState, action: Action): State {
+// Just mutating the substate of the tracker
+function updateUserState (state: TrackerState, action: Action): State | Object {
```
Comment (MarcoPolo) [inline react-native/react/reducers/tracker.js]: Why does this return an object, why can't it always return state?

File: react-native/react/reducers/tracker.js [hunk 9]
```diff
@@ -29,64 +28,64 @@ type State = {
...
-    followsYou: false,
-    avatar: null,
-    location: '' // TODO: get this information
+  trackers: {}
+}
+
+function initialTrackerState (): TrackerState {
+  return {
+    serverActive: false,
+    username: null,
```
Comment (MarcoPolo) [inline react-native/react/reducers/tracker.js]: this function should take a username

File: react-native/react/reducers/tracker.js [hunk 10]
```diff
@@ -16,8 +16,7 @@ import type {SimpleProofState, SimpleProofMeta} from '../constants/tracker'
 import type {Identity, RemoteProof, LinkCheckResult, ProofState, identifyUi_TrackDiffType, TrackSummary} from '../constants/types/flow-types'
 import type {Action} from '../constants/types/flux'
 
-type State = {
-  serverStarted: boolean,
+type TrackerState = {
   serverActive: boolean,
   proofState: SimpleProofState,
   username: ?string,
```
Comment (MarcoPolo) [inline react-native/react/reducers/tracker.js]: If the function below takes a username, then this is no longer a maybe string!

File: react-native/react/reducers/tracker.js (hunk 8 above)
Comment (chrisnojima) [inline react-native/react/reducers/tracker.js]: done

File: react-native/react/actions/tracker.js [hunk 11]
```diff
@@ -165,20 +187,21 @@ function serverCallMap (dispatch: Dispatch): CallMap {
...
         payload: {
-          active: false
+          active: false,
+          username
         }
       })
     }
   }
 
+  /*
```
Comment (MarcoPolo) [inline react-native/react/actions/tracker.js]: why are these commented out?

**Comment by MarcoPolo** [comment]
could we have the dev-menu tracker listener work as well?

I'd be okay with picking a random user or even just hardcoding the prop to be `username='max'`
but I think it's pretty useful for debugging (especially for the supposed ipc lag)

**Comment by MarcoPolo** [comment]
can we put a TODO somewhere to remember to unify different session id's about the same username?

I'm afraid that maybe two things will request a popup and expect a outcome (say track or not) and they'll get out of sync because only one of them will get replied to?

File: react-native/react/native/remote-manager.desktop.js [hunk 12]
```diff
@@ -34,29 +34,57 @@ class RemoteManager extends Component {
     }
   }
 
-  componentWillReceiveProps (nextProps) {
-    if (!this.props.trackerServerActive && nextProps.trackerServerActive) {
-      this.setState({showTrackerPopup: true})
-    }
+  windowStates (trackers) {
+    return Object.keys(trackers).map(user => {
```
Comment (MarcoPolo) [inline react-native/react/native/remote-manager.desktop.js]: Is order guaranteed for Object.keys?

File: react-native/react/reducers/tracker.js [hunk 13]
```diff
@@ -29,64 +28,67 @@ type State = {
...
+      followingCount: -1,
+      followsYou: false,
+      avatar: null,
+      location: '' // TODO: get this information
+    }
   }
 }
 
-export default function (state: State = initialState, action: Action): State {
+// Just mutating the substate of the tracker
```
Comment (MarcoPolo) [inline react-native/react/reducers/tracker.js]: this doesn't mutate

**Comment by MarcoPolo** [comment]
LGTM % comment

File: react-native/react/actions/tracker.js [hunk 14]
```diff
@@ -165,38 +186,26 @@ function serverCallMap (dispatch: Dispatch): CallMap {
...
         payload: {
-          active: false
+          active: false,
+          username
         }
       })
     }
   }
 
-  dispatch({
```
Comment (MarcoPolo) [inline react-native/react/actions/tracker.js]: why are these deleted? we need these guys

Comment (chrisnojima) [inline react-native/react/actions/tracker.js]: well the flow is kinda different as at this point we don't have a username or substore. so i can do that logic in the start: or really maybe we should just have that logic in the reducer init since it'll always be the same value on start right?

Comment (chrisnojima) [inline react-native/react/actions/tracker.js]: i'll just emit them when we setup teh session id, if thats wrong let me know

File: react-native/react/native/remote-manager.desktop.js (hunk 12 above)
Comment (chrisnojima) [inline react-native/react/native/remote-manager.desktop.js]: in es6 yes

**Comment by chrisnojima** [comment]
added todo: https://keybase.atlassian.net/browse/DESKTOP-108

**Comment by chrisnojima** [comment]
@MarcoPolo i'm going ot merge this since i addressed yoru comments but if something's not cool let's ticket and i'll fix it
